url: Fix error type in constructor

benjamingr opened the pull request:
Currently whatwg URLs fail with an incorrect error when null is passed as the base. Adding a check before accessing a symbol for the URL makes the URL error correctly.

Add test for it.

Fixes: https://github.com/nodejs/node/issues/19254

cc @targos @nodejs/url

benjamingr commented:
About semverness - I feel like this is a semver-patch since it's a bugfig. _Technically_ it changes an error message in an edge case.

targos commented on test/parallel/test-whatwg-url-parsing.js:
```diff
@@ -22,6 +22,7 @@ const failureTests = tests.filter((test) => test.failure).concat([
   { input: null },
   { input: new Date() },
   { input: new RegExp() },
+  { input: 'test', base: null},
```
Sorry, my example in the issue was not very good. This throws because `'test'` is not a valid URL. With this change, a valid input with a base of `null` will not throw. To make it work, we still need to reach the line with `base = new URL(base);`

benjamingr replied:
Thanks for clarifying!

benjamingr replied:
Added two relevant examples below

benjamingr replied:
@targos please take another look - I've changed the code since that review.

watilde commented on lib/internal/url.js:
```diff
@@ -312,6 +312,7 @@ class URL {
     // toUSVString is not needed.
     input = `${input}`;
     if (base !== undefined &&
+        base !== null &&
```
According to the current specification of URL, it might not be needed to checking base as like instanceof since base is guaranteed to be a USVString. Also, in this case, it should display an error for parsing `base = null`, so I think we can simply write like the following:
```js
if (base !== undefined) {
  base = new URL(base);
}
```

Refs: https://github.com/jsdom/whatwg-url/blob/master/lib/URL-impl.js#L12-L17
Spec: https://github.com/whatwg/url/blob/master/url.bs#L2592-L2599

benjamingr replied:
Yeah you're right, the undefined check is also incorrect

watilde replied:
Thanks for your quick update!

benjamingr commented:
@watilde @targos please take a look, I've added a test for a valid URL with invalid base and changed the implementation to not check if the object passed in is a URL and build a new URL anyway.

I didn't notice any regression compared to the optimization done in https://github.com/nodejs/node/pull/11690/files - I can readd the check here but I wonder how common the case a URL is passed in as a baseUrl (which becomes a little slower) is vs a string (which becomes a little faster, and is the more important case IMO). 

If either of you feels strongly about it I can re-add that check inside a `!== null`

targos commented on lib/internal/url.js:
```diff
@@ -311,8 +311,7 @@ class URL {
     input = `${input}`;
-    if (base !== undefined &&
-        (!base[searchParams] || !base[searchParams][searchParams])) {
+    if (arguments.length === 2) {
```
I think we should still allow to explicitly pass `undefined`. I don't know what the spec says, but Chrome accepts it.

benjamingr replied:
@targos https://github.com/whatwg/url/blob/master/url.bs#L2592-L2599 implies it doesn't allow it - but I'm all for Chrome compatibility and I'm not sure what the spec reads as. 

What do you think about just removing the `base[searchParams]` and staying compatible with the URL impl at https://github.com/jsdom/whatwg-url/blob/master/lib/URL-impl.js#L12-L17 ?

targos replied:
Staying compatible with whatwg-url SGTM. I would actually like us to throw a more precise error saying which URL us invalid but that's out of scope for this PR

watilde replied:
This is about how we can write "If base is given" and both sound good to me if CI gets green.

targos replied:
@TimothyGu What do you think about this?

watilde commented:
ci: https://ci.nodejs.org/job/node-test-pull-request/13650/

domenic commented:
To be clear, undefined is allowed, it's just then the spec steps for "base is given" do not run. (Undefined = not present, as always in JS.) So parsedBase gets left as null. In that case the first argument has to be an absolute URL, or the parser will fail (because that's what happens when you parse a relative URL against a null base).

benjamingr commented:
@domenic and what should the behavior for `null` be according to the spec? Should it attempt to create a base URL and raise an error?

domenic commented:
`null` is the same as `"null"`, which is not parseable as a base URL, so yeah.

benjamingr commented:
Ok, I basically ended up removing the optimization for passing an "already URL" as base (since I suspect that's not the interesting case anyway) - that removes the bug and I don't see a big performance difference.

targos commented on test/parallel/test-whatwg-url-parsing.js:
```diff
@@ -22,6 +22,9 @@ const failureTests = tests.filter((test) => test.failure).concat([
   { input: new RegExp() },
+  { input: 'test', base: null},
+  { input: 'http://nodejs.org', base: null},
+  { input: 'http://nodejs.org', base: undefined},
```
Does this still error? It should not.

benjamingr replied:
Good point!

benjamingr replied:
Done, also squashed.

benjamingr commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/13654/console

watilde commented on test/parallel/test-whatwg-url-parsing.js:
```diff
@@ -22,6 +22,8 @@ const failureTests = tests.filter((test) => test.failure).concat([
   { input: null },
   { input: new Date() },
   { input: new RegExp() },
+  { input: 'test', base: null},
```
nits: space is required before `}`. Also on line 26.

benjamingr replied:
Very concerned that the linter didn't fail locally - thanks!

TimothyGu commented:
@benjamingr would you be interested in submitting a pull request to [Web Platform Tests](https://github.com/w3c/web-platform-tests/blob/master/url/url-constructor.html) as well, to ensure that the same bug doesn't happen in browsers?

TimothyGu commented:
> Ok, I basically ended up removing the optimization for passing an "already URL" as base (since I suspect that's not the interesting case anyway) - that removes the bug and I don't see a big performance difference.

Just wondering, how did you measure the performance difference? I don't believe we have a benchmark for this case yet.

targos commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/14011/

BridgeAR commented:
Landed in cc6abc6e84b96fd5f1c4123066eba93ddb637e60 :tada:
